packages/grafana-ui/src/components/Table/TableNG/utils.ts

```diff
@@ -71,58 +81,117 @@ export function shouldTextWrap(field: Field): boolean {
   return Boolean(cellOptions?.wrapText);
 }
 
-// matches characters which CSS
-const spaceRegex = /[\s-]/;
+/**
+ * @internal creates a typography context based on a font size and family. used to measure text
+ * and estimate size of text in cells.
+ */
+export function createTypographyContext(fontSize: number, fontFamily: string, letterSpacing = 0.15): TypographyCtx {
+  const font = `${fontSize}px ${fontFamily}`;
+  const canvas = document.createElement('canvas');
+  const ctx = canvas.getContext('2d')!;
+
+  ctx.letterSpacing = `${letterSpacing}px`;
+  ctx.font = font;
+  const txt =
+    "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s";
+  const txtWidth = ctx.measureText(txt).width;
+  const avgCharWidth = txtWidth / txt.length + letterSpacing;
+  const { count } = varPreLine(ctx);
+
+  return {
+    count,
+    ctx,
+    font,
+    avgCharWidth,
+  };
+}
+
+/**
+ * @internal returns counts the number of lines in the value of a text cell using the typography context's count function.
+ */
+export function getTextLineCounter(typographyCtx: TypographyCtx): LineCounter {
+  return (value, width) => {
+    if (!value) {
+      return 1;
+    }
+    return typographyCtx.count(String(value), width);
+  };
+}
+
+/**
+ * @internal return a text line counter for every field which has wrapHeaderText enabled.
+ */
+export function buildHeaderLineCounters(fields: Field[], typographyCtx: TypographyCtx): LineCounterEntry[] {
+  const wrappedColIdxs = fields.reduce((acc, field, idx) => {
+    if (field.config?.custom?.wrapHeaderText) {
+      acc.push(idx);
+    }
+    return acc;
+  }, [] as number[]);
+
+  if (wrappedColIdxs.length === 0) {
+    return [];
+  }
+
+  return [{ counter: getTextLineCounter(typographyCtx), fieldIdxs: wrappedColIdxs }];
+}
+
+/**
+ * @internal return a text line counter for every field which has wrapHeaderText enabled.
+ */
+export function buildRowLineCounters(fields: Field[], typographyCtx: TypographyCtx): LineCounterEntry[] {
+  const result: Record<string, LineCounterEntry> = {};
+
+  for (let fieldIdx = 0; fieldIdx < fields.length; fieldIdx++) {
+    const field = fields[fieldIdx];
+    if (shouldTextWrap(field)) {
+      // TODO: Pills, DataLinks, and JSON will have custom line counters here.
+      if (field.type === FieldType.string) {
```

Comment:
this fixes the issue you saw with uwrap running too often @leeoniya - only string fields with `wrapText` set to true will get the text wrapper now.